Add remote authentication module

ubipo opened the pull request:
This adds a module for remote authentication, as described in #7.

I also upgraded the Maven version and targeted Java version because Android studio asked for it. Hope that doesn't break anything ;)

Since the `ModuleUserData` for the new authentication module does not need a granted phones list I moved that functionality to a new class, `PhoneAllowlistModuleUserData`. So all modules use `PhoneAllowlistModuleUserData`, except for the authentication module, which uses `ModuleUserData`. 

The parameter description for the authentication module got a little hectic, so I enabled HTML for that textview and updated the parameter descriptions for the other modules.

I'm probably missing some stuff but overall I tried to keep the changes small.

ubipo committed:
Bump Java and Gradle version

ubipo committed:
gitignore directorypicker build folder

ubipo committed:
Add initial remote authentication module

ubipo committed:
Implement auth module phone whitelist granting, don't show phone number whitelist for modules that don't need it

ubipo committed:
Remove idea assetWizardSettings

ubipo committed:
Update module descriptions

ubipo committed:
Remove stray logging statement

tranquvis commented on app/src/main/java/tranquvis/simplesmsremote/CommandManagement/Commands/CommandAuth.java:
```diff
@@ -0,0 +1,156 @@
+        }
+
+        // Fall back to locale specific title
+        for (Module module : modules) {
```
👍

tranquvis reviewed (changes requested):
Hi Pieter,

thanks for this great contribution. In summary, the implementation is of good quality and I also appreciate the changes, which don't affect the new feature. 
I hope you still have time to take a look on my comments and make these small changes.

Automated tests should also be added. You can take a look on the existing ones, it's not a big effort.

ubipo committed:
Implement suggestions (see #44)

ubipo committed:
Implemented instrumentation tests for GrantPhoneRemotely module

ubipo committed:
Remove some stray imports

ubipo commented on app/src/main/java/tranquvis/simplesmsremote/Data/AuthenticationModuleSettingsData.java:
```diff
@@ -0,0 +1,28 @@
+public class AuthenticationModuleSettingsData extends ModuleSettingsData {
+    private String password;
+
+    public static final String DEFAULT_PASSWORD = "password";
```
Good idea, I went for 6 chars capitals and numbers excluding "O" and "0".

ubipo commented on app/src/main/java/tranquvis/simplesmsremote/CommandManagement/Modules/ModuleAuthentication.java:
```diff
@@ -0,0 +1,23 @@
+    public final CommandAuth commandAuth = new CommandAuth(this);
+
+    public ModuleAuthentication() {
+        this.titleRes = R.string.control_module_title_authentication;
```
Ya, true. I used "Grant Phone Remotely" as the module name. Bit long, but I did want to show that it's remote management and not just managing granted phones on your phone somehow (although that wouldn't make sense as an sms module, but still). Also didn't include the "manage" part because you can - currently - only grant new phones, not manage old ones.

ubipo commented:
Thanks for the thorough review!

I went over all the suggestions and implemented the instrumentation tests. As with the module itself I did need to make some changes to the deeper logic.

I got 19 out of 60 failures but these all seem to be some functional problem (WifiManager.setWifiApEnabled not found, something with the display staying on when the test tried to turn it off etc.). The tests I implemented for GrantPhoneRemotely do all pass.

tranquvis commented on app/src/androidTest/java/tranquvis/simplesmsremote/CommandManagement/Commands/CommandGrantPhoneRemotelyTest.java:
```diff
@@ -0,0 +1,53 @@
+    @Test
+    public void testExecution() throws Exception {
+        assertThat("grant pwd all").matches().executes();
+        assertThat("grant pwd audio wifi-hotspot").matches().executes();
```
This does only successfully execute if the modules are already enabled. We should ensure that before executing the command.

tranquvis reviewed (approved):
Hi,

thanks for implementing my suggestions so fast and this great work 🥇  I'm not really satisfied with the test implementation (with mine too), but this shouldn't delay merging this PR.
